### PR TITLE
fix(compiler): could not hydrate optionals

### PR DESCRIPTION
**Pull request by hasanaburayyan**
There are several open issues in progress that need to be able to define generic optionals. 
Such as: 
- https://github.com/winglang/wing/issues/1868
- https://github.com/winglang/wing/issues/3575
- https://github.com/winglang/wing/issues/3653

I only added `tryAt` method for Array since the actual work is defined in another issue, but I needed an example to write some test cases with. Since there are several efforts that need this change I opted to not wait and include it in one of those PRs but just add the support here so other efforts can rebase.

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [x] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

**Comment by hasanaburayyan** [comment]
CC: @revitalbarletz

File: libs/wingc/src/type_check.rs [hunk 1]
```diff
@@ -3932,28 +3931,38 @@ impl<'a> TypeChecker<'a> {
...
-						WINGSDK_MUT_SET => self.types.add_type(Type::MutSet(t1_replacement)),
-						_ => {
-							self.unspanned_error(format!("\"{}\" is not a supported generic return type", fqn));
-							self.types.error()
+
+			match *type_to_maybe_replace {
+				Type::Optional(t) => {
+					let concrete_t = self.get_concrete_type_for_generic(t, types_map);
+					return self.types.add_type(Type::Optional(concrete_t));
+				}
```
Comment (MarkMcCulloh) [inline libs/wingc/src/type_check.rs:3939]: nit: to remove the extra nesting I would just forgo the extra match add this before the original code:

```rust
if let Type::Optional(t) = *type_to_maybe_replace {
  let concrete_t = self.get_concrete_type_for_generic(t, types_map);
  return self.types.add_type(Type::Optional(concrete_t));
}
```

File: libs/wingsdk/src/std/array.ts [hunk 2]
```diff
@@ -38,6 +38,19 @@ export class Array {
...
+   *
+   * @macro ($self$[$args$])
+   *
+   * @param index index of the value to get
+   * @returns the value at the given index, or undefined if the index is out of bounds
+   */
+  public tryAt(index: number): T1 | undefined {
+    index;
+    throw new Error("Macro");
+  }
```
Comment (MarkMcCulloh) [inline libs/wingsdk/src/std/array.ts:52]: Technically the current at() already does this, we just don't have it typed correctly. I guess we should add a macro to the current `at` to throw if undefined instead.

File: libs/wingsdk/src/std/array.ts [hunk 3]
```diff
@@ -38,6 +38,19 @@ export class Array {
     throw new Error("Abstract");
   }
 
+  /**
+   * Try to get the value at the given index
+   *
+   * @macro ($self$[$args$])
```
Comment (MarkMcCulloh) [inline libs/wingsdk/src/std/array.ts:44]: to handle negative indexes? idk I guess this stuff is out of scope for the focus of this PR.

```suggestion
   * @macro ($self$.at($args$))
```

File: libs/wingsdk/src/std/array.ts (hunk 2 above)
Comment (hasanaburayyan) [reply libs/wingsdk/src/std/array.ts:52]: This is included in the description for https://github.com/winglang/wing/issues/1868 which is already being worked on (not by me)

File: libs/wingc/src/type_check.rs (hunk 1 above)
Comment (hasanaburayyan) [reply libs/wingc/src/type_check.rs]: I constantly forget that you can check for a single case this way.

File: libs/wingsdk/src/std/array.ts [hunk 4]
```diff
@@ -38,6 +38,19 @@ export class Array {
     throw new Error("Abstract");
   }
 
+  /**
+   * Try to get the value at the given index
```
Comment (Chriscbr) [inline libs/wingsdk/src/std/array.ts:42]:
```suggestion
   * Get the value at the given index, returning nil if the index is out of bounds.
```

File: libs/wingsdk/src/std/array.ts
Comment (Chriscbr) [inline libs/wingsdk/src/std/array.ts:1]: Shouldn't the regular `array.at` function throw a runtime error if the index is out of bounds? The method says it returns a `T` so failing silently doesn't seem right.

**Review by MarkMcCulloh — APPROVED** [review]
